### PR TITLE
feat(linear): Add ensemble tree model 

### PR DESCRIPTION
**Pull request by shenkha**
## What does this PR do?

This pull request introduces a major enhancement to the linear tree-based models by implementing **Ensemble Tree Models**, which improves prediction accuracy and robustness over a single tree model.

#### Key Changes:

**1. Ensemble of Trees**
- A new `EnsembleTreeModel` class in `libmultilabel/linear/tree.py` now manages multiple tree models.
- The `train_ensemble_tree` function handles the training of `n` separate tree models, each with a different random seed for diversity.
- The ensemble's final predictions are an average of the scores from each tree, providing a more stable and accurate result.
- This functionality is exposed via a new CLI argument `--tree_ensemble_models` in `main.py` and integrated into `linear_trainer.py`.

*Example usage:*
```bash
python main.py --training_file data/eurlex_raw_texts_train.txt \
                --test_file data/eurlex_raw_texts_test.txt \
                --linear \
                --linear_technique tree \
                --tree_ensemble_models 3
```

## Test CLI & API (`bash tests/autotest.sh`)
Test APIs used by main.py.
- [x] Test Pass
  - (Copy and paste the last outputted line here.)
- [ ] Not Applicable (i.e., the PR does not include API changes.)

## Check API Document
If any new APIs are added, please check if the description of the APIs is added to API document. 
- [ ] API document is updated ([linear](https://www.csie.ntu.edu.tw/~cjlin/libmultilabel/api/linear.html), [nn](https://www.csie.ntu.edu.tw/~cjlin/libmultilabel/api/nn.html))
- [x] Not Applicable (i.e., the PR does not include API changes.)

## Test quickstart & API (`bash tests/docs/test_changed_document.sh`)
If any APIs in quickstarts or tutorials are modified, please run this test to check if the current examples can run correctly after the modified APIs are released.

### CODE REVIEW

File: libmultilabel/linear/tree.py [hunk 1]
```diff
@@ -6,12 +6,14 @@
 import scipy.sparse as sparse
 import sklearn.cluster
 import sklearn.preprocessing
+import sklearn.utils
 from tqdm import tqdm
 import psutil
-
 from . import linear
+from scipy.special import log_expit
+#from sparsekmeans import LloydKmeans, ElkanKmeans
```
Comment (will945945945) [inline libmultilabel/linear/tree.py]: line 14 should be removed

File: libmultilabel/linear/tree.py [hunk 2]
```diff
@@ -6,12 +6,14 @@
 import scipy.sparse as sparse
 import sklearn.cluster
 import sklearn.preprocessing
+import sklearn.utils
 from tqdm import tqdm
 import psutil
-
```
Comment (will945945945) [inline libmultilabel/linear/tree.py:11]: why we need to modify this line?

File: libmultilabel/linear/tree.py [hunk 3]
```diff
@@ -109,17 +130,14 @@ def _separate_model_for_pruning_tree(self):
                 **tree_flat_model_params
             )
             self.subtree_models.append(subtree_flatmodel)
-        
+
```
Comment (will945945945) [inline libmultilabel/linear/tree.py]: why we need to modify this line?

File: libmultilabel/linear/tree.py [hunk 4]
```diff
@@ -109,17 +130,14 @@ def _separate_model_for_pruning_tree(self):
...
     def _prune_tree_and_predict_values(self, x: sparse.csr_matrix, beam_width: int) -> np.ndarray:
         """Calculates the selective decision values associated with instances x by evaluating only the most relevant subtrees.
-
```
Comment (will945945945) [inline libmultilabel/linear/tree.py]: why we need to modify this line?

File: libmultilabel/linear/tree.py [hunk 5]
```diff
@@ -109,17 +130,14 @@ def _separate_model_for_pruning_tree(self):
...
         Only subtrees corresponding to the top beam_width candidates from the root are evaluated,
         skipping the rest to avoid unnecessary computation.
-
```
Comment (will945945945) [inline libmultilabel/linear/tree.py]: why we need to modify this line?

File: libmultilabel/linear/tree.py [hunk 6]
```diff
@@ -109,17 +130,14 @@ def _separate_model_for_pruning_tree(self):
...
         Args:
             x (sparse.csr_matrix): A matrix with dimension number of instances * number of features.
             beam_width (int): Number of top candidate branches considered for prediction.
-
```
Comment (will945945945) [inline libmultilabel/linear/tree.py]: why we need to modify this line?

File: libmultilabel/linear/tree.py [hunk 7]
```diff
@@ -179,18 +199,18 @@ def _beam_search(self, instance_preds: np.ndarray, beam_width: int) -> np.ndarra
...
-                children_score = score - np.square(np.maximum(0, 1 - pred))
+                children_score = self._get_scores(pred, score)
                 next_level.extend(zip(node.children, children_score.tolist()))
 
             cur_level = sorted(next_level, key=lambda pair: -pair[1])[:beam_width]
             next_level = []
 
         num_labels = len(self.root.label_map)
-        scores = np.zeros(num_labels)
+        scores = np.full(num_labels, 0.0)
```
Comment (will945945945) [inline libmultilabel/linear/tree.py]: why we need to modify this line?

File: libmultilabel/linear/tree.py [hunk 8]
```diff
@@ -47,20 +49,39 @@ def __init__(
...
+        options = self.options or ""
+        options_split = options.split()
+        if "-s" in options_split:
+            i = options_split.index("-s")
+            if i + 1 < len(options_split):
+                solver_type = options_split[i + 1]
+                return solver_type in ["0", "6", "7"]
+        return False
+
+    def _get_scores(self, pred, parent_score=0.0):
```
Comment (will945945945) [inline libmultilabel/linear/tree.py]: We should specify the parameter type. Please see other functions.

File: libmultilabel/linear/tree.py [hunk 9]
```diff
@@ -47,20 +49,39 @@ def __init__(
...
+        if self._is_lr():
+            #return parent_score - np.log(1 + np.exp(-pred))
```
Comment (will945945945) [inline libmultilabel/linear/tree.py]: line 74 should be removed.

File: libmultilabel/linear/tree.py (hunk 7 above)
Comment (shenkha) [reply libmultilabel/linear/tree.py]: my mistake, I have just checked and will revert right away

**Review by Eleven1Liu — COMMENTED** [review]
For the formatting issues mentioned above, please use [black formatter](https://pypi.org/project/black/).

File: linear_trainer.py [hunk 10]
```diff
@@ -21,7 +22,7 @@ def linear_test(config, model, datasets, label_mapping):
         scores = []
 
     predict_kwargs = {}
-    if model.name == "tree":
+    if model.name == "tree" or model.name == "ensemble-tree":
```
Comment (will945945945) [inline linear_trainer.py]: This can be modified to 

```python
if isinstance(model, (TreeModel, EnsembleTreeModel)):
```

The classes should be imported in line 9.

**Review by will945945945 — APPROVED** [review]
